docs/equality.md

```diff
@@ -0,0 +1,348 @@
+
+### C# or F# enum type
+
+* Semantics: User expects identical to equality on underlying
+* Perf: User expects same perf as flattening to underlying type
+* Compilation today: flattens to underlying type
+* Perf today: good ✅
+* [sharplab for C# enum int](https://sharplab.io/#v2:DYLgZgzgNALiCWwA+BYAUMApjABGHAFAB4g4DKAnhDJgLYB0AIgIYUDyYA6ppgNYCUOCjgC8hIqKH90QA===)
+* [sharplab for F# enum int](https://sharplab.io/#v2:DYLgZgzgNALiCWwA+BYAUDAngBwKYAIBRfAXn3X0qXwEFT8BGCq/AIXoCZ11hcZ8w+ABQAPEEQCU+TPVH1ME9EA=)
+
+### C# struct type
+
+* Semantics: User really expects call to `IEquatable<T>` if present, but F# spec says call `this.Equals(box that)`, in practice these are the same
+* Compilation today: `GenericEqualityIntrinsic<SomeStructType>`
+* Perf today: always boxes (❌, Problem1)
+* [sharplab](https://sharplab.io/#v2:DYLgZgzgNALiCWwA+BYAUMApjABGHAFAB4g4DKAnhDJgLYB0AIgIY0Aq8tmA8mJNgEocFHAF5CRMcIHogA==)
+
+### F# struct type (with compiler-generated structural equality)
+
+* Semantics: User expects field-by-field structural equality with no boxing
+* Compilation today: `GenericEqualityIntrinsic<SomeStructType>`
+* Perf today: always boxes (❌, Problem1b)
+* [sharplab](https://sharplab.io/#v2:DYLgZgzgNALiCWwA+BYAUAbQDwGUYCcBXAYxgD4BddGATwAcBTAAhwHsBbBvI0gCgDcQTeADsYUJoSGiYASiYBedExVNO7AEYN8TAPoA6AGqKm/ZavVadBgKonC6dMAYwmYJrwAeQtp24k5JhoTLxMaWXQgA)
+
+Note: the optimization path is a bit strange here, the reductions are:
+
+```fsharp
+
+(x = y) 
+
+--inline--> 
+
+GenericEquality x y 
+
+--inline--> 
+
+GenericEqualityFast x y 
+
+--inline--> 
+
+GenericEqualityIntrinsic x y
+
+--devirtualize-->
+
+x.Equals(box y, LanguagePrimitives.GenericEqualityComparer);
+```
+
+The struct type has these generated methods:
+```csharp
+    override bool Equals(object y) 
+    override bool Equals(SomeStruct obj)
+    override bool Equals(object obj, IEqualityComparer comp) //withcEqualsVal
+```
+
+These call each other in sequence, boing then bunboxing then boxing. We do NOT generate this method, we probably should:
+
+```csharp
+    override bool Equals(SomeStruct obj, IEqualityComparer comp) //withcEqualsValUnboxed
+```
+
+If we did, the devirtualizing optimization should reduce to this directly, which would result in no boxing.]
+
+### F# large ref record/union type
+
+Here "large" means the compiler-generated structural equality is not inlined
+
+* Semantics: User expects structural by default
+* Perf: User expects perf is sum of constituent parts, type-specialized if generic
+* Today: direct call to `Equals(T)`, which has specialized code but boxes fields if struct or T, see Problem1, Problem2 ❌
+
+### F# tiny ref record/union type
+
+Here "tiny" means the compiler-generated structural equality IS inlined
+
+* Semantics: User expects structural by default
+* Perf: User expects perf is sum of constituent parts, type-specialized if generic
+* Test: Equals06.fsx
+* Compilation Today: flattened, calling `GenericEqualityERIntrinsic` on struct and generic fields
```

Comment:
Since this doc goes to some very basics at places and does mention function names with `ER` or `PER` function names,  this should be explained as well.

(incl. the tradeoff it brings)

Reply:
Good point. I added an explanation.